src/content/docs/recipes/testing/snapshots.md

```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+// Now in tests module:
+#[cfg(test)]
+mod tests {
+    use super::App;
+    use insta::assert_snapshot;
+    use ratatui::{backend::TestBackend, Terminal};
+
+    #[test]
+    fn test_render_app() {
+        let app = App::default();
+        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
+        terminal
+            .draw(|frame| frame.render_widget(&app, frame.area()))
+            .unwrap();
+        assert_snapshot!(terminal.backend());
+    }
+}
+```
+
+### 3. Running the Test
+
+When you run the test (`cargo test`), the output of the `Terminal::draw()` method will be compared
+against a snapshot. If this is the first time running the test or the output has changed, [`insta`]
+will create a snapshot file under the `snapshots/` directory.
+
+For example, after running the test, a new snapshot file might be created at:
+
+```
+snapshots/demo2__tests__render_app.snap
+```
+
+This file will store the visual representation of your terminal as a string:
+
+```text
+---
+source: examples/demo2/main.rs
+expression: terminal.backend()
+---
+"Ratatui                               Recipe  Email  Traceroute  Weather        "
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▄▄███▄▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▄███████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▄█████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀████████████▄▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀                              ▀▀▀▀▀▀▀▀▀▀▀▀▀███████████▀▀▀▀▄▄██████▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  ──────── Ratatui ─────────  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀███▀▄█▀▀████████▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  - cooking up terminal user  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▄▄▄▄▀▄████████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  interfaces -                ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀████████████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀                              ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀███▀██████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  Ratatui is a Rust crate     ▀▀▀▀▀▀▀▀▀▀▀▀▀▄▀▀▄▀▀▀█████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  that provides widgets       ▀▀▀▀▀▀▀▀▀▀▀▄▀ ▄  ▀▄▀█████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  (e.g. Paragraph, Table)     ▀▀▀▀▀▀▀▀▀▄▀  ▀▀    ▀▄▀███████▄▄▄▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀  and draws them to the       ▀▀▀▀▀▀▀▄▀      ▄▄    ▀▄▀█████████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀                              ▀▀▀▀▀▄▀         ▀▀     ▀▄▀██▀▀▀███▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█                    ▀▄▀▀▀▄██▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▄                    ▀▄▀█▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀"
+"     H/←  Left  L/→  Right  K/↑  Up  J/↓  Down  D/Del  Destroy  Q/Esc  Quit     "
+```
+
+In the snapshot, each line represents a row of the terminal, capturing the rendered state of your
+TUI. The next time you run your tests, the output will be compared to this file to detect any
+unintentional changes.
```

Comment:
Yeah, that's because `Display` impl on `TestBackend` (i.e. [`buffer_view`](https://github.com/ratatui/ratatui/blob/0ed71d14bbe7d08905d525566f8c2678a5ffd57c/src/backend/test.rs#L49) method) does not take style into account. We can probably add another method for retrieving the colored buffer.

What is in your mind btw? Returning ANSI stuff or Debug strings?

Reply:
Not 100% sure - I'd guess that putting ANSI in the snapshot would likely make this look bad, but maybe it would be ok. The alternative is serialize to something that is easy to compare visually (e.g. `{red}foo{/red}`) or something like that. Either way, right now this can only test the text value of the output, which is good, but not excellent.

Reply:
9fef53f7d620c34ac213e1dff715b6a3859ac126

Reply:
It's not a huge issue, but out of curiosity, why does `backend.to_string()` put quotes around every line? It seems odd that the string conversion quotes every line of output, and makes the snapshots look a little weird.

Reply:
> why does backend.to_string() put quotes around every line?

It is the implementation detail of the `buffer_view` function:

https://github.com/ratatui/ratatui/blob/ee850159397385facd7b88e585b1331bca9ad2ef/src/backend/test.rs#L54-L54

Reply:
This was done to optimize for the TDD loop - write a broken test, see the debug view output in a form that works well in tests, copy the output directly into the test. Rerun the passing test. Effectively doing the same sort of review step you'd do with insta but manually.